libdevcore/MemoryDB.cpp

```diff
@@ -14,176 +14,95 @@
-        else
-            cwarn << "Lookup required for value with refcount == 0. This is probably a critical trie issue" << _h;
-    }
-    return std::string();
+    m_batch[_key.toString()] = _value.toString();
 }
 
-bool MemoryDB::exists(h256 const& _h) const
+void MemoryDBWriteBatch::kill(Slice _key)
 {
-#if DEV_GUARDED_DB
-    ReadGuard l(x_this);
-#endif
-    auto it = m_main.find(_h);
-    if (it != m_main.end() && (!m_enforceRefs || it->second.second > 0))
-        return true;
-    return false;
+    m_batch.erase(_key.toString());
 }
 
-void MemoryDB::insert(h256 const& _h, bytesConstRef _v)
+std::string MemoryDB::lookup(Slice _key) const
 {
-#if DEV_GUARDED_DB
-    WriteGuard l(x_this);
-#endif
-    auto it = m_main.find(_h);
-    if (it != m_main.end())
+    std::string value;
+    Guard lock(m_mutex);
+    auto const& it = m_db.find(_key.toString());
+    if (it != m_db.end())
     {
-        it->second.first = _v.toString();
-        it->second.second++;
+        value = it->second;
     }
-    else
-        m_main[_h] = make_pair(_v.toString(), 1);
-#if ETH_PARANOIA
-    cdebug << "INST" << _h << "=>" << m_main[_h].second;
-#endif
+    return value;
 }
 
-bool MemoryDB::kill(h256 const& _h)
+bool MemoryDB::exists(Slice _key) const
 {
-#if DEV_GUARDED_DB
-    ReadGuard l(x_this);
-#endif
-    if (m_main.count(_h))
-    {
-        if (m_main[_h].second > 0)
-        {
-            m_main[_h].second--;
-            return true;
-        }
-#if ETH_PARANOIA
-        else
-        {
-            // If we get to this point, then there was probably a node in the level DB which we need to remove and which we have previously
-            // used as part of the memory-based MemoryDB. Nothing to be worried about *as long as the node exists in the DB*.
-            cdebug << "NOKILL-WAS" << _h;
-        }
-        cdebug << "KILL" << _h << "=>" << m_main[_h].second;
-    }
-    else
-    {
-        cdebug << "NOKILL" << _h;
-#endif
-    }
-    return false;
+    Guard lock(m_mutex);
+    return m_db.count(_key.toString()) != 0;
 }
 
-bytes MemoryDB::lookupAux(h256 const& _h) const
+void MemoryDB::insert(Slice _key, Slice _value)
 {
-#if DEV_GUARDED_DB
-    ReadGuard l(x_this);
-#endif
-    auto it = m_aux.find(_h);
-    if (it != m_aux.end() && (!m_enforceRefs || it->second.second))
-        return it->second.first;
-    return bytes();
+    Guard lock(m_mutex);
+    m_db[_key.toString()] = _value.toString();
 }
 
-void MemoryDB::removeAux(h256 const& _h)
+void MemoryDB::kill(Slice _key)
 {
-#if DEV_GUARDED_DB
-    WriteGuard l(x_this);
-#endif
-    m_aux[_h].second = false;
+    Guard lock(m_mutex);
+    m_db.erase(_key.toString());
 }
 
-void MemoryDB::insertAux(h256 const& _h, bytesConstRef _v)
+std::unique_ptr<WriteBatchFace> MemoryDB::createWriteBatch() const
 {
-#if DEV_GUARDED_DB
-    WriteGuard l(x_this);
-#endif
-    m_aux[_h] = make_pair(_v.toBytes(), true);
+    return std::unique_ptr<WriteBatchFace>(new MemoryDBWriteBatch());
```

Comment:
Watch out with creating objects as `Object()`, compiler sometimes thinks it is a function definition (not in this context however). Always safe alternatives: `new Object`, `new Object{}`.

Reply:
Ah interesting, thanks for calling this out!